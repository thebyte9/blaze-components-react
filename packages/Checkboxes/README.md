## Description

Boxes that are checked (ticked) when activated. They allow you to select single values for submission in a form (or not).

## Usage

- Multiselect

```js
const multiple = [
  {
    label: "First",
    value: 1,
    id: "one"
  },
  {
    label: "Second",
    value: 2,
    id: "two"
  },
  {
    label: "Third",
    value: 3,
    id: "three"
  },
  {
    label: "Disabled",
    value: 4,
    disabled: true
  }
];

<Checkboxes options={multiple} onChange={({ event, value }) => {}} />;
```

- Single & required

```js
const single = {
  label: 'Do you agree?',
  required: true
};

<Checkboxes
  options={single}
  boolean
  onChange={({ event, value }) => {}}
```

## API

##### Checkboxes can receive a number of `props` as follow:

| NAME     |   TYPE   | DEFAULT  |
| :------- | :------: | :------: |
| options  |  Array   |    []    |
| disabled | Boolean  |  false   |
| required | Boolean  |  false   |
| boolean  | Boolean  |  false   |
| onChange | Function | () => {} |
