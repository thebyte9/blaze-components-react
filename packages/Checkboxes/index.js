import React, { Fragment, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import uuidv1 from 'uuid/v1';

const Checkboxes = ({ onChange, options, withEffect, boolean, ...attrs }) => {
  const [data, setData] = useState(Array.isArray(options) ? options : [options]);

  useEffect(
    () => {
      if (withEffect) setData(options);
    },
    [options, withEffect]
  );

  const toggle = ({ event, item, key }) => {
    if (item.disabled) return;
    data[key].checked = !item.checked;
    setData([...data]);
    let checked = data.filter(option => option.checked);
    if (boolean) checked = !!checked.length;
    onChange({ event, checked, data });
  };

  return (
    <Fragment>
      {data.map((item, key) => {
        const {
          checked = false,
          value,
          disabled,
          required,
          label,
          show = true,
          id = uuidv1()
        } = item;

        if (!show) return <Fragment key={id} />;

        return (
          <div
            key={id}
            className={`form-field form-field--checkbox ${required ? 'required' : ''}`}
            onClick={event => toggle({ event, item, key })}
            role="button">
            <input
              readOnly
              type="checkbox"
              className="form-checkbox"
              value={value}
              disabled={disabled}
              checked={checked}
              required={required}
              id={id}
              {...attrs}
            />
            <label htmlFor={id}>{label}</label>
          </div>
        );
      })}
    </Fragment>
  );
};

Checkboxes.propTypes = {
  options: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  withEffect: PropTypes.bool,
  boolean: PropTypes.bool,
  onChange: PropTypes.func
};

Checkboxes.defaultProps = {
  options: [],
  withEffect: false,
  boolean: false,
  onChange: () => {}
};

export default Checkboxes;
