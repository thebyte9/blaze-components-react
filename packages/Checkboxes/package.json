{
  "name": "@blaze-react/checkboxes",
  "version": "0.8.0-alpha.38",
  "description": "Boxes that are checked (ticked) when activated. They allow you to select single values for submission in a form (or not).",
  "main": "./dist/main.js",
  "scripts": {
    "build": "NODE_ENV=production webpack --config webpack.config.js",
    "build-storybook": "build-storybook -c .storybook -o docs-build",
    "lint": "eslint '*/**/*.{ts,tsx}' --ignore-path ../../.eslintignore --fix",
    "test": "jest --passWithNoTests",
    "test:coverage": "jest --coverage --passWithNoTests",
    "pre-publish": "yarn build && yarn lint && yarn test",
    "bundle-analyser": "node ../../node_modules/webpack-bundle-analyzer/lib/bin/analyzer.js ./stats/stats.json"
  },
  "keywords": [],
  "author": "byte9",
  "license": "GPL-3.0",
  "devDependencies": {
    "@babel/cli": "^7.13.14",
    "@babel/core": "^7.13.14",
    "@babel/plugin-proposal-class-properties": "^7.13.0",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "webpack": "^5.28.0",
    "webpack-cli": "^4.6.0"
  },
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "gitHead": "4bfe5dcb677b6f46229be4039ba4ca7db86e7cca"
}
