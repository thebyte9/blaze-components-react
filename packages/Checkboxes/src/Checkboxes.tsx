import ErrorMessage from '../../Utils/src/ErrorMessage';
import React, { FunctionComponent, useEffect, useState } from "react";
import Checkbox from "./Checkbox";
import { v4 as uuidv4 } from 'uuid';

interface IErrorMessage {
  message: string | JSX.Element;
  icon?: string;
}

export interface ICheckboxItem {
  checked: boolean;
  id: string | number;
  value: any;
}

interface ICheckBoxesProps {
  options?: any[] | Record<string, unknown>;
  returnBoolean?: boolean;
  onChange: ({
    event,
    value,
    data,
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    value: ICheckboxItem[];
    data: Record<string, unknown>[];
  }) => void;
  error?: boolean;
  name?: string;
  validationMessage?: string | JSX.Element;
  full?: boolean;
}

export { IErrorMessage, ICheckBoxesProps };

const CheckBoxes: FunctionComponent<ICheckBoxesProps> = ({
  returnBoolean,
  onChange: onChangeCheckboxList,
  options,
  error,
  validationMessage,
  name,
  full,
  ...attrs
}): JSX.Element => {
  const formatedOptions = Array.isArray(options) ? options : [options];

  const [data, setData]: any = useState(formatedOptions);

  useEffect(() => setData(formatedOptions), [options]);

  const onChange = ({
    value,
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    value: any;
  }): void => {
    if (value.disabled) {
      return;
    }

    const currentIndex = data.findIndex(
      ({ id }: { id: string }) => id === value.id
    );
    if (currentIndex < 0) {
      return;
    }

    data[currentIndex].checked = value.checked;
    setData([...data]);

    let currentValues = data.filter(
      ({ checked }: { checked: boolean }): boolean => checked
    );

    if (returnBoolean) {
      currentValues = !!currentValues.length;
    }

    onChangeCheckboxList({
      data,
      event: { target: { name } },
      value: currentValues,
    } as any);
  };

  return (
    <>
      {data.map(
        (
          {
            checked = false,
            disabled,
            id,
            label,
            name: checkboxName,
            required,
            show = true,
            value,
          }: any,
          key: number
        ): JSX.Element | null =>
          show ? (
            <Checkbox
              full={full}
              key={id || uuidv4()}
              onChange={onChange}
              checked={checked}
              disabled={disabled}
              id={id}
              label={label}
              name={checkboxName}
              required={required}
              show={show}
              value={value}
            />
          ) : null
      )}
      {error && <ErrorMessage message={validationMessage} />}
    </>
  );
};

CheckBoxes.defaultProps = {
  error: false,
  full: false,
  options: [],
  returnBoolean: false,
  validationMessage: "This field is required",
};

export default CheckBoxes;