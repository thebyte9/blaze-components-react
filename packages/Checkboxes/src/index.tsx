import withUtils from "@blaze-react/utils";
import React, { Fragment, FunctionComponent, useEffect, useState } from "react";
interface IErrorMessage {
  message: string | JSX.Element;
  icon?: string;
}

interface ICheckboxesProps {
  options?: any[] | object;
  returnBoolean?: boolean;
  onChange: ({
    event,
    value,
    data
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    value: boolean | object;
    data: object[];
  }) => void;
  error?: boolean;
  validationMessage: string | JSX.Element;
  utils: {
    uniqueId: (element: any) => string;
    classNames: (className: string | object, classNames?: object) => string;
    ErrorMessage: FunctionComponent<IErrorMessage>;
  };
}
const Checkboxes: FunctionComponent<ICheckboxesProps> = ({
  returnBoolean,
  onChange,
  options,
  error,
  validationMessage,
  utils: { uniqueId, classNames, ErrorMessage },
  ...attrs
}): JSX.Element => {
  const formatedOptions = Array.isArray(options) ? options : [options];

  const [data, setData]: any = useState(formatedOptions);

  useEffect(() => setData(formatedOptions), [options]);

  const toggle = ({
    event,
    item,
    key
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    item: any;
    key: number;
  }): void => {
    if (item.disabled) {
      return;
    }

    data[key].checked = !item.checked;
    setData([...data]);

    let value = data.filter(
      ({ checked }: { checked: boolean }): boolean => checked
    );

    if (returnBoolean) {
      value = !!value.length;
    }

    onChange({ event, value, data });
  };

  return (
    <Fragment>
      {data.map(
        (item: any, key: number): JSX.Element => {
          const {
            checked = false,
            value,
            disabled,
            required,
            label,
            show = true,
            id = uniqueId(item)
          } = item;

          if (!show) {
            return <Fragment key={id} />;
          }

          const checkboxClassName = classNames(
            "form-field form-field--checkbox",
            { required }
          );

          return (
            <div
              data-testid={`checkbox-${key + 1}`}
              key={id}
              className={checkboxClassName}
              onClick={(event): void => toggle({ event, item, key })}
              role="button"
            >
              <input
                readOnly
                type="checkbox"
                className="form-checkbox"
                value={value}
                disabled={disabled}
                checked={checked}
                required={required}
                id={id}
                {...attrs}
              />
              <label htmlFor={id}>{label}</label>
            </div>
          );
        }
      )}
      {error && <ErrorMessage message={validationMessage} />}
    </Fragment>
  );
};
Checkboxes.defaultProps = {
  error: false,
  options: [],
  returnBoolean: false,
  validationMessage: "This field is required"
};
export default withUtils(Checkboxes);
