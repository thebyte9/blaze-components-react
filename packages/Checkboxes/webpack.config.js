const path = require("path");
const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");
const ESLintPlugin = require("eslint-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');

module.exports = {
  mode: "production",
  entry: "./src/index.ts",
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: "[name].[contenthash].js",
    clean: true
  },
  optimization: {
    minimize: true,
    minimizer: [`...`, new CssMinimizerPlugin()],
    usedExports: true,
    removeAvailableModules: true,
    concatenateModules: true
  },
  resolve: {
    extensions: [".js", ".json", ".ts", ".tsx"],
  },
  module: {
    rules: [
      {
        test: /\.(ts|js)x?$/i,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: [
              "@babel/preset-env",
              "@babel/preset-react",
              "@babel/preset-typescript",
            ],
          },
        },
      },
      {
        test: /\.scss$/,
        use: [
          MiniCssExtractPlugin.loader,
          "css-loader",
          {
            loader: "postcss-loader",
            options: {
              postcssOptions: {
                plugins: [
                  ["postcss-preset-env"],
                ],
              },
            },
          },
          "sass-loader"
        ],
      }],
  },
  externals: {
    react: 'react',
    'react-dom': 'react-dom'
  },
  plugins: [
    new ForkTsCheckerWebpackPlugin({
      async: false,
    }),
    new ESLintPlugin({
      extensions: ["js", "jsx", "ts", "tsx"],
    }),
    new MiniCssExtractPlugin({
      filename: "[name].[contenthash].css",
    }),
  ],
};

