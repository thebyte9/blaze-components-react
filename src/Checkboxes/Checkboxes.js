import React, { Fragment, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import uuidv1 from 'uuid/v1';

const Checkboxes = ({
  onChange,
  options,
  withEffect,
  boolean,
  ...attrs
}) => {
  const [data, setData] = useState(Array.isArray(options) ? options : [options]);

  useEffect(() => {
    if (withEffect) setData(options);
  }, [options]);

  const toggle = ({ event, item, key }) => {
    if (item.disabled) return;
    data[key].checked = !item.checked;
    setData([...data]);
    let checked = data.filter(option => option.checked);
    if (boolean) checked = !!checked.length;
    onChange({ event, checked });
  };

  return (
    <Fragment>
      {data.map((item, key) => {
        const {
          checked = false,
          value,
          disabled,
          required,
          label,
          id = uuidv1()
        } = item;
      
        return (
          <span
                key={id}
                className="form-field form-field--checkbox"
                onClick={e => toggle({ e, item, key })}
                role="button">
            <input
                readOnly
                type="checkbox"
                className="form-checkbox"
                value={value}
                disabled={disabled}
                checked={checked}
                required={required}
                id={id}
                {...attrs}
                />
            &nbsp; &nbsp;
            <label
                htmlFor={id}
                className={required ? 'required' : ''}>
              {label}
            </label>
          </span>
        );
      })}
    </Fragment>
  );
};

Checkboxes.propTypes = {
  options: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object
  ]),
  withEffect: PropTypes.bool,
  boolean: PropTypes.bool,
  onChange: PropTypes.func
};

Checkboxes.defaultProps = {
  options: [],
  withEffect: false,
  boolean: false,
  onChange: () => {}
};

export default Checkboxes;
