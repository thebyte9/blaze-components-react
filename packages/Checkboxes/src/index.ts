import Checkboxes, { ICheckboxItem } from './Checkboxes';

import Checkbox from './Checkbox';

export { Checkbox, Checkboxes, ICheckboxItem };
