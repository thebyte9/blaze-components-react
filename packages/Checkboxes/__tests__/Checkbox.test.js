import React from 'react';
import expect from 'expect';
import { shallow } from 'enzyme';
import Checkboxes from '../src/index';

const options = [
  { label: 'Example', value: 1, id: 'one' },
  {
    label: 'I accept',
    value: 'accepted',
    required: true,
    id: 'two'
  },
  {
    label: 'Disabled',
    value: '',
    disabled: true,
    id: 'three'
  }
];

describe('Checkboxes component', () => {
  test('should be defined and renders correctly (snapshot)', () => {
    const wrapper = shallow(<Checkboxes options={options} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test('should toggle checkbox on click', () => {
    const wrapper = shallow(<Checkboxes boolean withEffect options={options} />);

    expect(
      wrapper
        .find('input')
        .at(1)
        .prop('checked')
    ).toBe(false);

    wrapper
      .find('.form-field')
      .at(1)
      .simulate('click');

    expect(
      wrapper
        .find('input')
        .at(1)
        .prop('checked')
    ).toBe(true);
  });

  test("can't interact when checkbox is disabled", () => {
    const wrapper = shallow(<Checkboxes options={options} />);

    expect(
      wrapper
        .find('input')
        .at(2)
        .prop('disabled')
    ).toBe(true);

    wrapper
      .find('.form-field')
      .at(2)
      .simulate('click');

    expect(
      wrapper
        .find('input')
        .at(2)
        .prop('checked')
    ).toBe(false);
    expect(
      wrapper
        .find('input')
        .at(2)
        .prop('disabled')
    ).toBe(true);
  });
});
