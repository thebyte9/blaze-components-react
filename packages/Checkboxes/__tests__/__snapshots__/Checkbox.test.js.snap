// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes component should be defined and renders correctly (snapshot) 1`] = `ShallowWrapper {}`;
